misc/python/materialize/checks/scenarios_upgrade.py

```diff
@@ -90,6 +90,60 @@ def actions(self) -> List[Action]:
         ]
 
 
+class UpgradeEntireMzSkipVersion(Scenario):
+    """Upgrade the entire Mz instance from the previous version directly to the current HEAD"""
+
+    def base_version(self) -> MzVersion:
+        return previous_version
+
+    def actions(self) -> List[Action]:
+        print(f"Upgrading starting from tag {self.base_version()} directly to HEAD")
+        return [
+            # Start with previous_version
+            StartMz(tag=previous_version),
+            Initialize(self),
+            Manipulate(self, phase=1),
+            # Upgrade directly to current source
+            KillMz(),
+            StartMz(tag=None),
+            Manipulate(self, phase=2),
+            Validate(self),
+            # A second restart while already on the current source
+            KillMz(),
+            StartMz(tag=None),
+            Validate(self),
+        ]
+
+
+class UpgradeEntireMzFourVersions(Scenario):
+    """Test upgrade X-4 -> X-3 -> X-2 -> X-1 -> X"""
+
+    def base_version(self) -> MzVersion:
+        return released_versions[3]
+
+    def actions(self) -> List[Action]:
+        print(f"Upgrading going through {released_versions[:3]}")
+        return [
+            StartMz(tag=released_versions[3]),
+            Initialize(self),
+            KillMz(),
+            StartMz(tag=released_versions[2]),
+            Manipulate(self, phase=1),
+            KillMz(),
+            StartMz(tag=released_versions[1]),
+            Manipulate(self, phase=2),
+            KillMz(),
+            StartMz(tag=released_versions[0]),
+            Validate(self),
```

Comment:
Would it make sense to have more manipulate-phases here? Maybe we could define it so that a manipulate(x) can run as long as initialize and manipulate(x-1) have been run. But probably not worth the effort if it's just for this test.

Reply:
I had a divine intervention tonight l and will try to formalize the entire upgrade problem with diagrams as such in Notion. We can continue the discussion there.

If we allowed more "manipulate" phases, someone needs to write them, at least for some Checks.